File: packages/react-native-sdk/src/message-template.ts
export interface FluxyMessageTemplate {
  id: string;
  name: string;
  body: string;
  vars?: string[];
  createdByUserId?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface FluxySendMessageOptions {
  templateId?: string;
  templateVars?: Record<string, string | number | boolean | null | undefined>;
  expiresInSeconds?: number;
  expiresAt?: string;
  visibility?: 'room' | 'whisper' | `role:${string}`;
  visibleTo?: string[];
}

export type FluxyPresenceIntent = 'chat' | 'compose' | 'idle';

export interface FluxyProjectActivity {
  id: string;
  kind: string;
  roomId?: string;
  userId?: string;
  data?: unknown;
  createdAt: string;
}

export function renderMessageTemplate(template: FluxyMessageTemplate, vars: Record<string, string | number | boolean | null | undefined>): string {
  let out = template.body;
  for (const [key, value] of Object.entries(vars)) {
    out = out.replace(new RegExp(`\\{\\{\\s*${key}\\s*\\}\\}`, 'g'), String(value ?? ''));
  }
  return out;
}

export function extractTemplateVarNames(body: string): string[] {
  const names: string[] = [];
  let i = 0;
  while (i < body.length) {
    const open = body.indexOf('{{', i);
    if (open === -1) break;
    const close = body.indexOf('}}', open + 2);
    if (close === -1) break;
    const inner = body.slice(open + 2, close).trim();
    if (/^\w+$/.test(inner)) names.push(inner);
    i = close + 2;
  }
  return [...new Set(names)];
}

File: packages/flutter-sdk/lib/src/message_template.dart
class MessageTemplate {
  final String id;
  final String name;
  final String body;
  final List<String> vars;
  final String? createdByUserId;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const MessageTemplate({required this.id, required this.name, required this.body, this.vars = const [], this.createdByUserId, this.createdAt, this.updatedAt});

  factory MessageTemplate.fromJson(Map<String, dynamic> json) {
    return MessageTemplate(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      body: json['body'] ?? '',
      vars: json['vars'] != null ? List<String>.from(json['vars']) : [],
      createdByUserId: json['createdByUserId'],
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.tryParse(json['updatedAt']) : null,
    );
  }
}

String renderMessageTemplate(MessageTemplate template, Map<String, dynamic> vars) {
  var out = template.body;
  for (final entry in vars.entries) {
    out = out.replaceAll(RegExp('\\{\\{\\s*${entry.key}\\s*\\}\\}'), entry.value?.toString() ?? '');
  }
  return out;
}

List<String> extractTemplateVarNames(String body) {
  final matches = RegExp(r'\{\{\s*(\w+)\s*\}\}').allMatches(body);
  return matches.map((m) => m.group(1)!).toSet().toList();
}
